include/CL/sycl/vec.hpp

```diff
@@ -63,7 +63,7 @@ class vec : public detail::small_array<DataType,
 
 
   // Inherit of all the constructors
-  using typename basic_type::small_array;
+  using basic_type::small_array;
```

Comment:
This I understand, but after Dániel and I investigating the case, this very much seems to be a Clang bug. The following sample code

```C++
#include <iostream>

struct Q{ int i; };

namespace detail
{
template<typename T, typename U> struct base
{
  Q q;
  base( Q const& qq ):q(qq){ printf("q"); }
};
}
template <typename T> struct vec : public detail::base<T, vec<T>>
{
    //using Z = typename detail::base<T, vec<T>>;
 using detail::base<T, vec<T>>::base;
};

int main()
{
  Q q{2};
  vec<float> v( q );
}
```
We checked with Godbolt ([GCC 6.3](https://godbolt.org/g/yvnfU7), [Clang 3.9](https://godbolt.org/g/hUzlzW)) and the [online nightly](http://webcompiler.cloudapp.net/) MSVC (no link to copy+pasted code). GCC compiles both versions (strange/liberal approach), Clang fails with the alias/typedef proxy and MSVC compiles it. I am not a die hard language expert, but I still think that Clang is wrong. basic_type cannot be anything else, other than a typename, because it preceeds the :: , hence typename refers to the entire statement, and the CTOR is not a typename but a function.

Because basic_type appears only twice in class vec, I'd vote for typing it out twice (only 1 extra if omitting the alias) until someone issues a bug report to Clang and it is fixed. Without typename we believe it is conforming code, but we'd argue if it's valid with it. We can always ask the gurus on [StackOverflow](http://stackoverflow.com/questions/41938309/inheriting-constructor-and-template-aliase).